wannierberri/__Data_K.py

```diff
@@ -670,6 +673,34 @@ def S_H(self):
     @lazy_property.LazyProperty
     def S_H_rediag(self):
         return np.einsum("knna->kna",self.S_H).real
+#PRB RPS19, Ryoo's way to calculate SHC
+
+    @lazy_property.LazyProperty
+    def SA_H(self):
+        return self._R_to_k_H(self.SA_R.copy(), hermitian=False)
+    
+    @lazy_property.LazyProperty
+    def SHA_H(self):
+        return self._R_to_k_H(self.SHA_R.copy(), hermitian=False)
+#PRB QZYZ18, Qiao's way to calculate SHC
+
+    @lazy_property.LazyProperty
+    def SH_H(self):
+        return self._R_to_k_H(self.SH_R.copy(), hermitian=False)
+
+    @lazy_property.LazyProperty
+    def K_H(self):
+        return -1j*self._R_to_k_H(self.SR_R.copy(), hermitian=False) + np.einsum('knlc,klma->knmac',self.S_H,self.D_H)
```

Comment:
I check the time consumed, and the times with qiao and without qiao were similar. So I decided not to use an alternative.

Also, I tried to use the similar thing as kubo_sum_elements in kubo.py, but I failed to compute faster. My computer even stopped when I used

    def mat_product_knmac(self, x, y):
        #To compute np.einsum('knlc,klma->knmac', x, y) faster
        assert x.shape == (x.shape[0], self.num_wann, self.num_wann, 3)
        assert y.shape == (y.shape[0], self.num_wann, self.num_wann, 3)
        x_reshape = x.reshape((x.shape[0] * self.num_wann * 3, self.num_wann))
        y_reshape = y.reshape((self.num_wann, y.shape[0] * self.num_wann * 3))

        return (x_reshape @ y_reshape).reshape((-1, self.num_wann, self.num_wann, 3, 3))

to calculate K_H and L_H. I want to know the reason.

Reply:
Ok, let's leave it like it is for now.

Reply:
fwiw, I installed the current `wannierberri` for a user that is testing this new SHC feature and I can confirm that at least in his case, `np.einsum` is definitely the bottleneck, the job spends almost all of its time there

Reply:
Dear @migueldiascosta, if possible, could you share the input script and data that the user has? You or the user may send the data to my email jaemolim96@gmail.com.
I may try to optimize this part together with Minsu @manxkim (the author of this PR).

Reply:
I asked the user to get in touch with you. I'm pretty sure the matrices are quite large, since each process was using ~20GB of memory, so it may indeed take large matrices to see `np.einsum` becoming a bottleneck

Reply:
More generically, it may be worth looking into (if you haven't already) `np.einsum`'s `optimize` parameter, and the associated `einsum_path` for saving the optimal strategy for each geometry, which among other things checks if it is better to use a BLAS operation (`_can_dot`) (https://github.com/numpy/numpy/blob/2dbfb458fa3e96203f93118a9d142a6bedca9271/numpy/core/einsumfunc.py)

(Recent versions of numpy try to optimize by default, but they don't cache the best strategy for repeated calls to the same geometry, that needs to be done explicitly when relevant)

However, in this particular SHC case even then `_can_dot` returns False in condition 
```python
        # can't do implicit summation or dimension collapse e.g.                                                                                                                              
        #     "ab,bc->c" (implicitly sum over 'a')                                                                                                                                            
        #     "ab,ca->ca" (take diagonal of 'a')                                                                                                                                              
        if nl + nr - 1 == int(c in result):
            return False
```
when `c` is 'k', but maybe it is worth factoring out the 'k' index from the `einsum` call?

my two (sing)cents, fwiw